mibidata/tiff.py

```diff
@@ -285,47 +288,47 @@ def _page_metadata(page, description):
+    _convert_from_previous(description)
+
+    metadata = {}
+    for key, val in description.items():
+        if key.startswith('mibi.'):
+            metadata[key[5:]] = val
+        elif key not in RESERVED_MIBITIFF_ATTRIBUTES:
+            metadata[key] = val
 
-    # retrieve required metadata
-    metadata = {key.split('.')[1]: value for key, value in
-                description.items() if key.startswith('mibi.')}
     metadata.update({
         'coordinates': (
             _cm_to_micron(page.tags['x_position'].value),
             _cm_to_micron(page.tags['y_position'].value)),
         'date': date,
         'size': size})
-    # retrieve user-defined metadata
-    try:
-        user_defined_attrs = description['user_defined_attributes']
-        print(user_defined_attrs)
-        metadata.update({key: value for key, value in description.items()
-                         if key in user_defined_attrs})
-    except KeyError:
-        # no user-defined metadata found
-        pass
 
     return metadata
 
 
-def _convert_from_previous_metadata_versions(description):
-    """Convert old metadata format to new one.
+def _convert_from_previous(description):
+    """Convert old metadata format for backwards compatibility.
 
-    This function ensures backwards compatibility for previous versions.
+    Most of these conversions would happen during MibiImage construction,
+    but we do them here in case reading the info only.
     """
-    try:
-        description['mibi.MIBItiff_version']
-    except KeyError:
-        # if the key doesn't exist it means that the version was None and hence
-        # not saved in the tiff file
-        description['mibi.MIBItiff_version'] = None
-
-    if description['mibi.MIBItiff_version'] != mi.MIBITIFF_VERSION:
-        description['mibi.fov_name'] = description['mibi.description']
+    if not description.get('mibi.fov_name') and description.get(
+            'mibi.description'):
+        description['mibi.fov_name'] = description.pop('mibi.description')
+    # TODO: Clean up repetition between this and the same MibiImage method
+    if description.get('mibi.folder') and not description.get('mibi.fov_id'):
         description['mibi.fov_id'] = description['mibi.folder'].split('/')[0]
-        description['mibi.description'] = None
-        description['mibi.MIBItiff_version'] = mi.MIBITIFF_VERSION
+        warnings.warn(
+            'The "fov_id" attribute is now required if "folder" is '
+            'specified. Setting "fov_id" to {}.'.format(
+                description['mibi.fov_id']))
+    if (not description.get('mibi.folder') and description.get('mibi.fov_id')
+            and description.get('mibi.fov_id').startswith('FOV')):
+        description['mibi.folder'] = description['mibi.fov_id']
+        warnings.warn(
+            'The "folder" attribute is required if "fov_id" is specified. '
+            'Setting "folder" to {}.'.format(description['mibi.folder']))
 
```

Comment:
This logic is a bit complicated to follow. The previous implementation based on the MIBITIFF version was cleaner.

Reply:
I don't love this section of the code because it feels redundant to what happens during construction of a MibiImage, but relying on the user to say `MIBItiff_version=1.0` is less robust than looking at the actual parameters passed and converting them to the current version if necessary.